### PR TITLE
Determine causal window frames to produce early results.

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change
In streaming execution, window functions generate result when frame end is smaller than batch size (This guarantees that future rows may not extend up the current window frame). However, for causal window frame (e.g window frame that doesn't include future rows than the current row). We can generate results even if frame end is same with batch size. 

With this support we can decrease latency for causal queries.

## What changes are included in this PR?
This PR adds support for causality analysis for window frame to decrease latency and memory during generating window function results.

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/expr/src/window_frame.rs [hunk 1]
```diff
@@ -30,21 +30,24 @@ use sqlparser::ast;
...
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[derive(Clone, PartialEq, Eq, Hash)]
 pub struct WindowFrame {
     /// A frame type - either ROWS, RANGE or GROUPS
     pub units: WindowFrameUnits,
     /// A starting frame boundary
     pub start_bound: WindowFrameBound,
     /// An ending frame boundary
     pub end_bound: WindowFrameBound,
+    /// Flag indicates whether window frame is causal.
```
Comment (alamb) [inline datafusion/expr/src/window_frame.rs:49]: Can you please document what `causal` means in this context (I am not sure I understand what it means)? Perhaps you can update the doc comments on `WindowFrame` to explain what the concept is and give an example of a causal bound and non causal bound and then link to that definition when referred to elsewhere?

File: datafusion/sqllogictest/test_files/window.slt [hunk 2]
```diff
@@ -3871,3 +3871,31 @@ FROM (SELECT c1, c2, ROW_NUMBER() OVER(PARTITION BY c1) as rn
...
+          amount FLOAT
+        ) as VALUES
+          (0, '2022-01-01 06:00:00'::timestamp, 'EUR', 30.0),
+          (1, '2022-01-01 08:00:00'::timestamp, 'EUR', 50.0),
+          (2, '2022-01-01 11:30:00'::timestamp, 'TRY', 75.0),
+          (3, '2022-01-02 12:00:00'::timestamp, 'EUR', 200.0),
+          (4, '2022-01-03 10:00:00'::timestamp, 'TRY', 100.0)
+
+query TT
+EXPLAIN SELECT *, SUM(amount) OVER(ORDER BY sn) as sum1
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/window.slt:3890]: Could you add some comments here explaining what this test is validating (specifically so if it changes in the future we will have some additional context on the intent of the test)?

File: datafusion/sql/src/expr/function.rs [hunk 3]
```diff
@@ -102,6 +103,21 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 // Numeric literals in window function ORDER BY are treated as constants
                 false,
             )?;
+
+            let func_deps = schema.functional_dependencies();
```
Comment (alamb) [inline datafusion/sql/src/expr/function.rs:107]: Could you document what ordering is stable means ?

**Review by alamb — COMMENTED** [review]
Thank you for this PR @mustafasrepo -- I feel I lack the context to review it properly, maybe could you explain the usecase (ideally in doc comments somewhere) with an example or two? 

Also, the PR description talks about emitting batches earlier, but I didn't see any test coverage of that. Was that intended?

File: datafusion/sql/src/expr/function.rs (hunk 3 above)
Comment (mustafasrepo) [reply datafusion/sql/src/expr/function.rs:107]: Sure, updated comment

File: datafusion/sqllogictest/test_files/window.slt (hunk 2 above)
Comment (mustafasrepo) [reply datafusion/sqllogictest/test_files/window.slt:3898]: Thanks @alamb, added a comment to explain rationale

**Comment by mustafasrepo** [comment]
> Thank you for this PR @mustafasrepo -- I feel I lack the context to review it properly, maybe could you explain the usecase (ideally in doc comments somewhere) with an example or two?
> 
> Also, the PR description talks about emitting batches earlier, but I didn't see any test coverage of that. Was that intended?

 

> Add new test



> Thank you for this PR @mustafasrepo -- I feel I lack the context to review it properly, maybe could you explain the usecase (ideally in doc comments somewhere) with an example or two?
> 
> Also, the PR description talks about emitting batches earlier, but I didn't see any test coverage of that. Was that intended?

At first, I thought it would be hard to exactly test this behaviour. However, after some thinking I found out it wouldn't be that hard. I have added a test for this case in [commit](https://github.com/apache/arrow-datafusion/pull/8842/commits/4723dc687699787ebbeb950173c785ab7f39110b) Thanks for pointing this out.

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @mustafasrepo

**Review by ozankabak — APPROVED** [review]
I did another round of review after @alamb and it LGTM

**Comment by ozankabak** [comment]
I will go ahead and merge this since it is a fairly simple change that is an incremental improvement to existing functionality without any major change to code structure. In case of any issues related to this PR, let us know and we will address with a quick follow-on PR.

File: datafusion/expr/src/window_frame.rs [hunk 4]
```diff
@@ -23,32 +23,76 @@
...
-    /// A starting frame boundary
+    /// Starting frame boundary
     pub start_bound: WindowFrameBound,
-    /// An ending frame boundary
+    /// Ending frame boundary
     pub end_bound: WindowFrameBound,
-    /// Flag indicates whether window frame is causal
-    /// See documentation for [is_frame_causal] for what causal means in this context and how it is calculated.
-    is_causal: bool,
+    /// Flag indicating whether the frame is causal (i.e. computing the result
```
Comment (alamb) [inline datafusion/expr/src/window_frame.rs:49]: 👍

**Comment by alamb** [comment]
This slightly changed the API to create `WindowFrame` (can't use struct syntax to create anymore) so marking as API change. You can still create the struct using `WindowFrame::new_bounds`

**Pull request merged**